File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    return dfs(numbers, 0, target)

def dfs(numbers, i, target):
    if i == len(numbers):
        return sum(numbers) == target
    numbers1 = numbers[:i] + [numbers[i]] + numbers[i+1:]
    numbers2 = numbers[:i] + [-numbers[i]] + numbers[i+1:]
    return sum([dfs(numbers1, i+1, target), dfs(numbers2, i+1, target)])



# def dfs(numbers, idx, target):
#     if idx == len(numbers):
#         if sum(numbers) == target:
#             return 1
#         return 0
#     
#     numbers2 = numbers[:]
#     numbers2[idx] = -numbers[idx]
#     return dfs(numbers, idx+1, target) + dfs(numbers2, idx+1, target)
# 
# def solution(numbers, target):
#     answer = 0
#     answer = dfs(numbers, 0, target)
#     return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int dfs(vector<int>& nums, int depth, int now_sum, int target) {
	if (depth == nums.size()) {
        return now_sum == target;
    }
    
    return dfs(nums, depth + 1, now_sum + nums[depth], target)
    	+ dfs(nums, depth + 1, now_sum - nums[depth], target);
}

int solution(vector<int> numbers, int target) {
    return dfs(numbers, 0, 0, target);
}
